File: ProblemSolving/450DSA/Python/src/dynamicprogramming/LargestSumContiguousSubarray.py
"""
@author Anirudh Sharma

Given an array arr of N integers. Find the contiguous sub-array with maximum sum.
"""


def maxSubarraySum(a):
    # Length of the array
    n = len(a)
    # Global maxima
    globalMax = a[0]
    localMax = a[0]
    # Loop through the elements of the list
    for i in range(1, n):
        localMax = max(a[i], a[i] + localMax)
        globalMax = max(localMax, globalMax)
    return globalMax


if __name__ == "__main__":
    a = [1, 2, 3, -2, 5]
    print(maxSubarraySum(a))

    a = [-1, -2, -3, -4]
    print(maxSubarraySum(a))

File: ProblemSolving/450DSA/JavaScript/src/dynamicprogramming/LargestSumContiguousSubarray.js
/**
 * @author Anirudh Sharma
 * 
 * Given an array arr of N integers. Find the contiguous sub-array with maximum sum.
 */
const maxSubarraySum = (a) => {
    // Length of the array
    const n = a.length;
    // Global maxima
    let globalMax = a[0];
    let localMax = a[0];
    // Loop through the elements of the list
    for (let i = 1; i < n; i++) {
        localMax = Math.max(a[i], a[i] + localMax);
        globalMax = Math.max(localMax, globalMax);
    }
    return globalMax;
};

const main = () => {
    let a = [1, 2, 3, -2, 5];
    console.log(maxSubarraySum(a));

    a = [-1, -2, -3, -4];
    console.log(maxSubarraySum(a));
};

main();